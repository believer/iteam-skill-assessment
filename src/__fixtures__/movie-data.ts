import { build, fake } from 'test-data-bot'
import { SearchMovie, MovieType, Movie } from '../types'

type PartialMovie = Pick<
  Movie,
  'Actors' | 'Poster' | 'Title' | 'Type' | 'Year' | 'imdbID'
>

export const searchMovieBuilder = build<SearchMovie>('Movie').fields({
  Poster: fake(f => f.internet.url()),
  Title: fake(f => f.lorem.words()),
  Type: MovieType.Movie,
  Year: fake(f =>
    f.date
      .past()
      .getFullYear()
      .toString()
  ),
  imdbID: fake(f => f.random.alphaNumeric()),
})

export const movieBuilder = build<PartialMovie>('Movie').fields({
  Actors: 'test, test2, test3',
  Poster: fake(f => f.internet.url()),
  Title: fake(f => f.lorem.words()),
  Type: MovieType.Movie,
  Year: fake(f =>
    f.date
      .past()
      .getFullYear()
      .toString()
  ),
  imdbID: fake(f => f.random.alphaNumeric()),
})
