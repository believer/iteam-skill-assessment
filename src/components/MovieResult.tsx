import { Link } from '@reach/router'
import React from 'react'
import { LoadingBlock } from './LoadingBlock'
import { SearchMovie } from '../types'
import classNames from 'classnames'

interface MovieResultProps {
  movie: SearchMovie
}

export const Poster: React.FC<React.HTMLProps<HTMLImageElement> & {
  hover?: boolean
}> = ({ src, alt, hover = true }) => {
  return (
    <img
      src={src}
      alt={alt}
      className={classNames(
        'rounded shadow-lg object-fill md:h-80 w-full md:w-auto transition-transform transition-fast transition-ease-in-out',
        { 'hover:transform-up-2': hover }
      )}
    />
  )
}

const MovieResult: React.FC<MovieResultProps> = ({ movie }) => {
  const hasPoster = movie.Poster && movie.Poster !== 'N/A'

  return (
    <div className="rounded">
      <Link to={`/movie/${movie.imdbID}`}>
        {!hasPoster && <LoadingBlock className="h-80 w-full" />}
        {hasPoster && <Poster src={movie.Poster} />}
        <div className="text-center mt-4 font-bold text-gray-800">
          {movie.Title}
        </div>
      </Link>
    </div>
  )
}

export default MovieResult
