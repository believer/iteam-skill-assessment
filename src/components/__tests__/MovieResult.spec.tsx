import React from 'react'
import MovieResult, { Poster } from '../MovieResult'
import { render } from '@testing-library/react'
import { searchMovieBuilder } from '../../__fixtures__/movie-data'

test('it renders movies with missing poster', () => {
  const movie = searchMovieBuilder({
    Poster: 'N/A',
  })

  const { getByText, queryByAltText } = render(<MovieResult movie={movie} />)

  expect(getByText(movie.Title)).toBeInTheDocument()
  expect(queryByAltText(movie.Title)).not.toBeInTheDocument()
})

test('it renders poster without hover state', () => {
  const { container } = render(
    <Poster alt="test" hover={false} src="test.jpg" />
  )

  expect(container).toMatchSnapshot()
})
