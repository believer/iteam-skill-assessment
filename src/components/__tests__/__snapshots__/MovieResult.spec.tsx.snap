// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders poster without hover state 1`] = `
<div>
  <img
    alt="test"
    class="rounded shadow-lg object-fill md:h-80 w-full md:w-auto transition-transform transition-fast transition-ease-in-out"
    src="test.jpg"
  />
</div>
`;
